clang-tools-extra/clang-tidy/daedalean/OperatorOverloadingCheck.cpp

```diff
@@ -0,0 +1,213 @@
+      if (hasOperator(method->getParent(), OO_Subscript,
+                      function->getParamDecl(0)->getType(), true)) {
+        break;
+      }
+
+      diag(function->getBeginLoc(), "If operator[] overloaded with non-const version, const version MUST be present.") << function;
+    }
+    break;
+  case OO_EqualEqual:
+    checkOperatorPair(function, OO_ExclaimEqual, "If == overloaded, != MUST be overloaded as well");
+    if (hasOperator(function, OO_Less) && !hasOperator(function, OO_LessEqual)) {
+      diag(function->getBeginLoc(), "If == and < overloaded, <= MUST be overloaded as well") << function;
+    }
+    break;
+  case OO_ExclaimEqual:
+     checkOperatorPair(function, OO_EqualEqual, "If != overloaded, == MUST be overloaded as well");
+    break;
+  case OO_Less:
+    checkOperatorPair(function, OO_Greater, "If < overloaded, > MUST be overloaded as well");
+    break;
+  case OO_Greater:
+    checkOperatorPair(function, OO_Less, "If > overloaded, < MUST be overloaded as well");
+    break;
+  case OO_LessEqual:
+    checkOperatorPair(function, OO_GreaterEqual, "If <= overloaded, >= MUST be overloaded as well");
+    checkOperatorPair(function, OO_EqualEqual, "If <= overloaded, == MUST be overloaded as well");
+    break;
+  case OO_GreaterEqual:
+    checkOperatorPair(function, OO_LessEqual, "If >= overloaded, <= MUST be overloaded as well");
+    checkOperatorPair(function, OO_EqualEqual, "If >= overloaded, == MUST be overloaded as well");
+    break;
+  default:
+    break;
+  };
+}
+
+void OperatorOverloadingCheck::checkOperatorPair(const FunctionDecl * function, OverloadedOperatorKind kind, const std::string & message) {
+  if (hasOperator(function, kind)) {
+    return;
+  }
+
+  diag(function->getBeginLoc(), message) << function;
+}
+void OperatorOverloadingCheck::checkReturnSelf(const Stmt *stmt) {
+  if (!stmt) {
+    return;
+  }
+
+  if (const auto ptr = llvm::dyn_cast<ForStmt>(stmt); ptr) {
+    checkReturnSelf(ptr->getBody());
+  }
+  if (const auto ptr = llvm::dyn_cast<WhileStmt>(stmt); ptr) {
+    checkReturnSelf(ptr->getBody());
+  }
+  if (const auto ptr = llvm::dyn_cast<DoStmt>(stmt); ptr) {
+    checkReturnSelf(ptr->getBody());
+  }
+  if (const auto ptr = llvm::dyn_cast<IfStmt>(stmt); ptr) {
+    checkReturnSelf(ptr->getThen());
+    checkReturnSelf(ptr->getElse());
+  }
+  if (const auto ptr = llvm::dyn_cast<CompoundStmt>(stmt); ptr) {
+    for (const auto s : ptr->body()) {
+      checkReturnSelf(s);
+    }
+  }
+  if (const auto ptr = llvm::dyn_cast<ReturnStmt>(stmt); ptr) {
+    const auto expr = ptr->getRetValue();
+    if (const auto uOp = llvm::dyn_cast<UnaryOperator>(expr); uOp) {
+      if (uOp->getOpcode() != UO_Deref || !llvm::isa<CXXThisExpr>(uOp->getSubExpr())) {
+        diag(ptr->getBeginLoc(), "Assignment operator must return reference to this");
+      }
+    } else {
+      diag(ptr->getBeginLoc(), "Assignment operator must return reference to this");
+    }
+  }
+
```

Comment:
This function checks that if a return statement appears in the function, then it returns *this. I'm assuming you rely on the fact that if there is no return statement at all (or some path without a return statement), the compiler would raise a warning or an error for it already. If that is the case it would be good to add a comment here explaining it.

Reply:
added comment